controllers/propagator/template_utils.go

```diff
@@ -0,0 +1,645 @@
+		},
+	)
+	if err != nil {
+		resolverCtxCancel()
+
+		return nil, fmt.Errorf(
+			"failed to instantiate the template resolver for the service account %s: %w", serviceAccount, err,
+		)
+	}
+
+	rv := templateResolverWithCancel{
+		cancel:         resolverCtxCancel,
+		referenceCount: &atomic.Uint32{},
+	}
+
+	t.wg.Add(1)
+
+	go func() {
+		err := dynamicWatcher.Start(resolverCtx)
+
+		rv.canceled.Store(true)
+
+		if err != nil {
+			log.Error(err, "The DynamicWatcher for the service account unexpectedly stopped")
+
+			if err := t.onBackgroundError(ctx, serviceAccount); err != nil {
+				log.Error(err, "Failed to handle the DynamicWatcher unexpectedly stopping")
+			}
+		}
+
+		resolverCtxCancel()
+		t.wg.Done()
+	}()
+
+	<-dynamicWatcher.Started()
+
+	templateResolver, err := templates.NewResolverWithDynamicWatcher(dynamicWatcher, templates.Config{
+		AdditionalIndentation: 8,
+		DisabledFunctions:     []string{},
+		SkipBatchManagement:   true,
+		StartDelim:            TemplateStartDelim,
+		StopDelim:             TemplateStopDelim,
+	})
+	if err != nil {
+		log.Error(err, "Failed to instantiate a template resolver for the service account")
+
+		resolverCtxCancel()
+
+		return nil, err
+	}
+
+	rv.resolver = templateResolver
+
+	return &rv, nil
+}
+
+// onBackgroundError is called when a goroutine encounters an unrecoverable error. This will clean up the template
+// resolver and trigger reconciles on all replicated policies that leverage this template resolver.
+func (t *TemplateResolvers) onBackgroundError(ctx context.Context, serviceAccount types.NamespacedName) error {
+	replicatedPolicies := policiesv1.PolicyList{}
+
+	t.globalLock.Lock()
+
+	if t.TemplateResolvers[serviceAccount] != nil {
+		t.TemplateResolvers[serviceAccount].cancel()
+		t.TemplateResolvers[serviceAccount].canceled.Store(true)
+		t.TemplateResolvers[serviceAccount].resolver = nil
+	}
+
+	t.globalLock.Unlock()
+
+	// This is pulled from the controller-runtime cache which is why client-side filtering with filterFunc is ok.
+	err := t.mgrClient.List(ctx, &replicatedPolicies, &client.ListOptions{LabelSelector: replicatedPlcLabelSelector})
+	if err != nil {
+		return err
+	}
+
+	for i := range replicatedPolicies.Items {
+		replicatedPolicy := replicatedPolicies.Items[i]
+
+		if replicatedPolicy.Spec.HubTemplateOptions == nil {
+			continue
+		}
+
+		if replicatedPolicy.Spec.HubTemplateOptions.ServiceAccountName != serviceAccount.Name {
+			continue
+		}
+
+		_, rootNS, err := common.ParseRootPolicyLabel(replicatedPolicy.Name)
+		if err != nil {
+			continue
+		}
+
+		if rootNS != serviceAccount.Namespace {
+			continue
+		}
+
+		t.replicatedPolicyUpdates <- event.GenericEvent{Object: &replicatedPolicy}
+	}
+
+	return nil
+}
+
+// RemovePolicy will clean up watches on the current template resolver (service account used in the last call to
+// GetResolver) and if this was the last policy using this template resolver, the template resolver will be cleaned up.
```

Comment:
root policy or replicated policy? It is confusing. Base on referenceCount description it is a replicated policy right?

Reply:
Yup, it's a replicated policy. I can rename the method or clarify the description. Any preference?

Reply:
RemoveRePolicy? I don't know Dale is good at this lol